app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php

```diff
@@ -504,14 +505,17 @@ private function cleanupJobs($groupId, $currentTime)
         $connection = $scheduleResource->getConnection();
         $count = 0;
         foreach ($historyLifetimes as $status => $time) {
-            $count += $connection->delete(
-                $scheduleResource->getMainTable(),
-                [
-                    'status = ?' => $status,
-                    'job_code in (?)' => array_keys($jobs),
-                    'created_at < ?' => $connection->formatDate($currentTime - $time)
-                ]
-            );
+            // Range is not guaranteed to use index for each group and could deadlock, so ensure we use the job_code+status index
+            foreach (array_keys($jobs) as $jobCode) {
+                $count += $connection->delete(
+                    $scheduleResource->getMainTable(),
+                    [
+                        'job_code = ?' => $jobCode,
+                        'status = ?' => $status,
+                        'created_at < ?' => $connection->formatDate($currentTime - $time)
```

Comment:
I believe to fix problem better to replace `created_at` with `scheduled_at`  (see CRON_SCHEDULE_SCHEDULED_AT_STATUS index) and  it's  all
Also 2.3 branch have batch of improvements that can resolve problem, see https://github.com/magento/magento2/blob/2.3/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php

Reply:
That does not fix the issue. Already tested by adding new index for status and created_at.
Issue is as described where because the job_code is an array of values the index selection does not always select an index which separates the groups. So it is potential for one group to scan the rows belonging to another in order to locate the necessary rows to delete, and this is where deadlocks can occur as groups begin to scan each others rows.

This solution here I have confirmed resolved the issue in two environments. It's been running since I opened this PR and I have received no more deadlock errors.

Additionally, I see no changes in 2.3 branch that would resolve or impact the deadlock issue. I was getting deadlocks against 2.3.4.

Reply:
Any question let me know - it's been difficult to describe the behaviour of the indexes so I know it might not be fully clear 😬 Thanks for the input though it good to make sure I didn't miss anything.

Reply:
OK I just spotted the deadlock retrier.

I guess it's a choice between resolving the issue or keeping that retrier workaround, or both. I'm never a fan of circuit breaker or retry loops when there's an actual solution available. Maybe both combined would be useful if I rebase so that the retrier is only needed in exceptional circumstances.

Reply:
I do believe the current PR is a simpler approach too and the retrier just adds overheads and complexity to maintenance.

Reply:
Can I ask you to provide "explain" for queries before/after and version of MySQL you test with.  Because it's hard to believe that many queries works better then one (but some times it's )

Reply:
Also if you have some newrelic/munin/etc stats on this queries from production server will like to see it

Reply:
*Deletion*

Before patch:

```
explain delete FROM `cron_schedule` WHERE (status = 'success') AND (job_code in ('process_email_queue', 'update_reports', 'generate_pos_images', 'process_tryagain', 'migrate_order_payments')) AND (created_at < '2020-04-27 05:43:11');
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
| id | select_type | table         | partitions | type  | possible_keys          | key                    | key_len | ref   | rows | filtered | Extra       |
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
|  1 | DELETE      | cron_schedule | NULL       | range | CRON_SCHEDULE_JOB_CODE | CRON_SCHEDULE_JOB_CODE | 767     | const |  881 |   100.00 | Using where |
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
1 row in set (0.00 sec)
```

After adding index but still using IN for job_code:

```
mysql> explain delete FROM `cron_schedule` WHERE (status = 'success') AND (job_code in ('process_email_queue', 'update_reports', 'generate_pos_images', 'process_tryagain', 'migrate_order_payments')) AND (created_at < '2020-04-27 05:43:11');
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
| id | select_type | table         | partitions | type  | possible_keys                 | key                           | key_len | ref         | rows | filtered | Extra       |
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
|  1 | DELETE      | cron_schedule | NULL       | range | CRON_SCHEDULE_JOB_CODE_STATUS | CRON_SCHEDULE_JOB_CODE_STATUS | 790     | const,const |   69 |   100.00 | Using where |
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
1 row in set (0.00 sec)
```

After patch for a single job_code:

```
mysql> explain delete FROM `cron_schedule` WHERE (status = 'success') AND (job_code = 'migrate_order_payments') AND (created_at < '2020-04-27 05:43:11');
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
| id | select_type | table         | partitions | type  | possible_keys                 | key                           | key_len | ref         | rows | filtered | Extra       |
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
|  1 | DELETE      | cron_schedule | NULL       | range | CRON_SCHEDULE_JOB_CODE_STATUS | CRON_SCHEDULE_JOB_CODE_STATUS | 790     | const,const |   20 |   100.00 | Using where |
+----+-------------+---------------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------+
1 row in set (0.00 sec)
```

*Atomic update of status*

Before patch:

```
explain UPDATE `cron_schedule` AS `current`  LEFT JOIN `cron_schedule` AS `existing` ON existing.job_code = current.job_code AND (existing.executed_at > UTC_TIMESTAMP() - INTERVAL 1 DAY OR existing.executed_at IS NULL) AND existing.status = 'running' SET `current`.`status` = 'running' WHERE (current.schedule_id = '8093526') AND (current.status = 'pending') AND (existing.schedule_id IS NULL);
+----+-------------+----------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------------------+
| id | select_type | table    | partitions | type  | possible_keys          | key                    | key_len | ref   | rows | filtered | Extra                   |
+----+-------------+----------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------------------+
|  1 | UPDATE      | current  | NULL       | const | PRIMARY                | PRIMARY                | 4       | const |    1 |   100.00 | NULL                    |
|  1 | SIMPLE      | existing | NULL       | ref   | CRON_SCHEDULE_JOB_CODE | CRON_SCHEDULE_JOB_CODE | 767     | const |  428 |    10.00 | Using where; Not exists |
+----+-------------+----------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------------------+
2 rows in set (0.00 sec)
```

After patch:

```
explain UPDATE `cron_schedule` AS `current`  LEFT JOIN `cron_schedule` AS `existing` ON existing.job_code = current.job_code AND (existing.executed_at > UTC_TIMESTAMP() - INTERVAL 1 DAY OR existing.executed_at IS NULL) AND existing.status = 'running' SET `current`.`status` = 'running' WHERE (current.schedule_id = '8093507') AND (current.status = 'pending') AND (existing.schedule_id IS NULL);
+----+-------------+----------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------------------+
| id | select_type | table    | partitions | type  | possible_keys                 | key                           | key_len | ref         | rows | filtered | Extra                   |
+----+-------------+----------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------------------+
|  1 | UPDATE      | current  | NULL       | const | PRIMARY                       | PRIMARY                       | 4       | const       |    1 |   100.00 | NULL                    |
|  1 | SIMPLE      | existing | NULL       | ref   | CRON_SCHEDULE_JOB_CODE_STATUS | CRON_SCHEDULE_JOB_CODE_STATUS | 790     | const,const |    1 |    10.00 | Using where; Not exists |
+----+-------------+----------+------------+-------+-------------------------------+-------------------------------+---------+-------------+------+----------+-------------------------+
2 rows in set (0.00 sec)
```

Reply:
Just for deletion, replacing `created_at` with  `scheduled_at`   do all work
```
mysql> desc DELETE FROM `cron_schedule` WHERE (status = 'status') AND (job_code in ('indexer_reindex_all_invalid', 'indexer_update_all_views', 'indexer_clean_all_changelogs')) AND (created_at < '2020-04-26 12:44:06');
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
| id | select_type | table         | partitions | type  | possible_keys          | key                    | key_len | ref   | rows | filtered | Extra       |
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
|  1 | DELETE      | cron_schedule | NULL       | range | CRON_SCHEDULE_JOB_CODE | CRON_SCHEDULE_JOB_CODE | 767     | const |    9 |   100.00 | Using where |
+----+-------------+---------------+------------+-------+------------------------+------------------------+---------+-------+------+----------+-------------+
1 row in set (0.00 sec)

mysql> explain DELETE FROM `cron_schedule` WHERE (status = 'status') AND (job_code in ('indexer_reindex_all_invalid', 'indexer_update_all_views', 'indexer_clean_all_changelogs')) AND (scheduled_at < '2020-04-26 12:44:06');
+----+-------------+---------------+------------+-------+----------------------------------------------------------+-----------------------------------+---------+-------+------+----------+-------------+
| id | select_type | table         | partitions | type  | possible_keys                                            | key                               | key_len | ref   | rows | filtered | Extra       |
+----+-------------+---------------+------------+-------+----------------------------------------------------------+-----------------------------------+---------+-------+------+----------+-------------+
|  1 | DELETE      | cron_schedule | NULL       | range | CRON_SCHEDULE_JOB_CODE,CRON_SCHEDULE_SCHEDULED_AT_STATUS | CRON_SCHEDULE_SCHEDULED_AT_STATUS | 5       | const |    1 |   100.00 | Using where |
+----+-------------+---------------+------------+-------+----------------------------------------------------------+-----------------------------------+---------+-------+------+----------+-------------+
1 row in set (0.00 sec)
```